torchqdynamics/odeint.py

```diff
@@ -83,35 +90,24 @@ def _fixed_odeint(qsolver, y0, t_save, dt, exp_ops, save_states):
         y_save = torch.zeros(len(t_save), *y0.shape).to(y0)
 
     if len(exp_ops) > 0:
-        exp_save = torch.zeros(len(t_save), len(exp_ops)).to(
-            device=y0.get_device(), dtype=torch.float
-        )
-
-    # save first step
-    save_counter = 0
-    if t_save[0] == 0.0:
-        if save_states:
-            y_save[0] = y0
-        for j, op in enumerate(exp_ops):
-            exp_save[save_counter, j] = torch.trace(op @ y0)
-        save_counter += 1
+        exp_save = torch.zeros(len(exp_ops),
+                               len(t_save)).to(device=y0.device, dtype=torch.float)
```

Comment:
The `dtype` should not be `torch.float` by default. If you do cat qubits and look at bitflips typically, you want machine precision on the expectation value of the sigma_x invariant. Also, the `dtype` is not necessarely a real-valued element. If you evaluate the annihilation operator a for instance, it will be complex-valued. So I propose taking the same dtype as the `y0`.

Reply:
Right 👍